src/compiler/checker.ts

```diff
@@ -34228,6 +34340,434 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+            isPropertyDeclaration(node) ? createClassFieldDecoratorContextType(thisType, valueType) :
+            Debug.failBadSyntaxKind(node);
+        const overrideType = getClassMemberDecoratorContextOverrideType(nameType, isPrivate, isStatic);
+        return getIntersectionType([contextType, overrideType]);
+
+    }
+
+    function createClassAccessorDecoratorTargetType(thisType: Type, valueType: Type) {
+        return tryCreateTypeReference(getGlobalClassAccessorDecoratorTargetType(/*reportError*/ true), [thisType, valueType]);
+    }
+
+    function createClassAccessorDecoratorResultType(thisType: Type, inputType: Type, outputType: Type) {
+        return tryCreateTypeReference(getGlobalClassAccessorDecoratorResultType(/*reportError*/ true), [thisType, inputType, outputType]);
+    }
+
+    function createClassFieldDecoratorInitializerMutatorType(thisType: Type, inputType: Type, outputType: Type) {
+        const thisParam = createParameter("this" as __String, thisType);
+        const valueParam = createParameter("value" as __String, outputType);
+        return createFunctionType(/*typeParameters*/ undefined, thisParam, [valueParam], inputType, /*typePredicate*/ undefined, 1);
+    }
+
+    /**
+     * Creates a call signature for an ES Decorator. This method is used by the semantics of
+     * `getESDecoratorCallSignature`, which you should probably be using instead.
+     */
+    function createESDecoratorCallSignature(targetType: Type, contextType: Type, nonOptionalReturnType: Type) {
+        const targetParam = createParameter("target" as __String, targetType);
+        const contextParam = createParameter("context" as __String, contextType);
+        const returnType = getUnionType([nonOptionalReturnType, voidType]);
+        return createCallSignature(/*typeParameters*/ undefined, /*thisParameter*/ undefined, [targetParam, contextParam], returnType);
+    }
+
+    /**
+     * Gets a call signature that should be used when resolving `decorator` as a call. This does not use the value
+     * of the decorator itself, but instead uses the declaration on which it is placed along with its relative
+     * position amongst other decorators on the same declaration to determine the applicable signature. The
+     * resulting signature can be used for call resolution, inference, and contextual typing.
+     */
+    function getESDecoratorCallSignature(decorator: Decorator) {
+        // We are considering a future change that would allow the type of a decorator to affect the type of the
+        // class and its members, such as a `@Stringify` decorator changing the type of a `number` field to
+        // `string`, or a `@Callable` decorator adding a call signature to a `class`. For now, the type arguments
+        // for the various context types that will eventually change to reflect such mutations will be stubbed out
+        // with fixed types so that we have a convenient place to apply these mutations.

```

Comment:
I can remove some of the aliased variables, but I'd prefer to keep most of the comment in this section so that we have context as to the decorator application order for any future changes.